### ISSUE TITLE
Delp2 gives processor-dependence results 

### ISSUE THREAD

**Issue by JosephThomasParker** [boutproject/BOUT-dev#195]
In `master`, I find that `Delp2` evaluates to different things depending on the number of mpi tasks used.  Consequently example cases like `blob2d` and `hasagawa-wakatani` also depend on processor count.

`test-delp2` checks processor dependence, and currently passes because the difference is only ~10^-9.  But surely the calculation should be exact?  

I believe the source of the discrepancy is `ZFFT` around line 582 in `src/mesh/difops.cxx`, but rough testing (in `fix_delp2`) shows FFTs working as expected.  Can anyone shine some light on this?

**Comment by d7919** [comment]
I've noticed related things in the past (#58) but this was put down to differing floating round off due to operations being performed in a different order etc.

Is there a reason you suspect `ZFFT`? I'd have thought this would be one of the things that wouldn't change with processor count as `z` should always be local and each `x` and `y` is independent.

**Comment by JosephThomasParker** [comment]
I commented out the ZFFT and replaced its output with something that was processor-independent, and test-delp2 passes with a tolerance of 1e-128.  

FFTW's FAQs have this on FFT planning being non-deterministic:

> If you use FFTW_MEASURE or FFTW_PATIENT mode, then the algorithm FFTW employs is not deterministic: it depends on runtime performance measurements. This will cause the results to vary slightly from run to run. However, the differences should be slight, on the order of the floating-point precision, and therefore should have no practical impact on most applications.
>
>If you use saved plans (wisdom) or FFTW_ESTIMATE mode, however, then the algorithm is deterministic and the results should be identical between runs. 
> 
> http://www.fftw.org/faq/section3.html#nondeterministic  

However, the default in ZFFT is estimate.

Incidentally, fft_fftw.cxx also has the option of measure.  Setting this in the input file:

`[fft]`
`fft_measure = true`

the error is 1e-14.

**Comment by ZedThree** [comment]
Is this now resolved? Was the issue because `fft_measure` was used?

**Comment by d7919** [comment]
Propose to close as duplicate of #340

**Comment by JosephThomasParker** [comment]
No, it still happens in `master` for both `fft_measure` (which is non-deterministic) and `fft_estimate` (which is deterministic).  It might be due to the order of operations being different, as @d7919 says, though turning off compiler optimization didn't change the result.
